src/common/commonutils/CommandUtils.c

```diff
@@ -29,270 +31,194 @@ static int NormalizeStatus(int status)
-                KillProcess(timerProcess);
+                // It was a timeout, nothing to read, loop.
+                continue;
             }
-            else
+
+            bytesRead = read(pipefd[0], buffer, BUFFER_SIZE);
+            if (bytesRead == 0)
             {
-                // Timer process is done, kill the timed out worker process
-                OsConfigLogDebug(log, "Command timed out or it was canceled, command process killed (%d)", status);
-                KillProcess(workerProcess);
+                // Child closed the pipe, we are done.
+                status = 0;
+                break;
+            }
+            if (bytesRead < 0)
+            {
+                if (EINTR == errno) {
+                    continue;
+                }
+                OsConfigLogError(log, "Error reading from pipe for command '%s' - %s", command, strerror(errno));
+                status = -1;
+                break;
             }
 
-            // Wait on the remaining child (either worker or timer) and exit (the intermediate process)
-            waitpid(0, &intermediateStatus, 0);
-            _exit(status);
-        }
-        else if (intermediateProcess > 0)
-        {
-            // Wait on the intermediate process to finish
-            waitpid(intermediateProcess, &status, 0);
-        }
-        else
-        {
-            status = -1;
-            OsConfigLogError(log, "Failed forking intermediate process (errno: %d, '%s')", errno, strerror(errno));
-        }
-    }
-    else
-    {
-        OsConfigLogDebug(log, "SystemCommand: executing command '%s' without timeout or cancelation on %s thread", command, mainProcessThread ? "main process" : "worker");
-        if (0 == (workerProcess = fork()))
-        {
-            // Worker process
-            status = execl("/bin/sh", "sh", "-c", command, (char*)NULL);
-            _exit(status);
-        }
-        else if (workerProcess > 0)
-        {
-            // Wait on the worker process to terminate
-            waitpid(workerProcess, &status, 0);
-        }
-        else
-        {
-            // If our fork fails, try system(), if that also fails then the call fails
-            OsConfigLogError(log, "Failed forking process to execute command (errno: %d, '%s'), attempting system", errno, strerror(errno));
-            status = system(command);
-        }
-    }
-
-    status = NormalizeStatus(status);
-    OsConfigLogDebug(log, "SystemCommand: command '%s' returning %d", command, status);
-
-    return status;
-}
-
-#define MAX_COMMAND_RESULT_FILE_NAME 100
-
-int ExecuteCommand(void* context, const char* command, bool replaceEol, bool forJson, unsigned int maxTextResultBytes, unsigned int timeoutSeconds, char** textResult, CommandCallback callback, OsConfigLogHandle log)
-{
-    const char commandTextResultFileTemplate[] = "/tmp/~OSConfig.TextResult%u";
-    const char commandSeparator[] = " > ";
-    const char commandTerminator[] = " 2>&1";
```

Comment:
>const char commandTerminator[] = " 2>&1"; [](http://example.com/codeflow?start=4&length=41)

We needed this to capture error output from the executed command and return as text results. How are we achieving the same with the pipes?

Reply:
We don't need it as we're capturing both stderr and stdout with the pipe.

Reply:
Awesome!